design/one-pager-generic-usage-type.md

```diff
@@ -0,0 +1,397 @@
+deleted before their dependents. It’s possible to author a set of interdependent
+resources that will eventually create successfully, but may not eventually
+delete successfully.
+
+Two variants of this problem are:
+
+- **Orphaned Managed Resources (MRs).** Consider a Helm Release MR that is
+deployed to an EKS Cluster MR. The Helm Release has a dependency on the EKS
+Cluster’s connection details - it uses them as Provider credentials. If the EKS
+Cluster is deleted before the Helm Release the Helm Release controller becomes
+unable to connect to its Provider in order to delete the Release. In practice
+the Release External Resource (ER) was implicitly deleted along with the GKE
+Cluster - there’s actually no work left for the Helm release controller to do -
+but it cannot know that. This results in the Helm Release MR being ‘orphaned’
+in a pending deletion that can never be satisfied.
+
+- **Orphaned ‘Side-Effect’ Resources.** Consider the same example as above, and
+assume the Helm Release includes a Kubernetes Service resource of type:
+LoadBalancer. A side-effect of the creation of this Service will be to create an
+ELB in the EKS cluster’s VPC. This ELB is not managed by or even known to
+Crossplane, and is not deleted along with the EKS cluster. This means that the
+Helm Release (and therefore its Service) must be deleted before the EKS cluster
+in order to ensure the EKS cluster’s controllers trigger deletion of the ELB.
+If it is not, the ELB will be orphaned, costing money and blocking the deletion
+of the VPC it was created in.
+
+These problems are exacerbated by the fact that Kubernetes uses "background
+cascading deletion" by default. Put otherwise, the successful deletion of a
+parent resource is not dependent on the successful deletion of its children.
+Instead, the parent resource will be deleted successfully and the children
+(i.e. resources that declare the parent as an owner reference) will later be
+deleted by the Kubernetes garbage collector. This makes it difficult to diagnose
+or even notice resources that become orphaned when their parent XR is deleted.
+
+## Goals
+
+- Support "cross-provider" dependencies, for example an MR from provider-helm
+depending on a ProviderConfig from provider-aws.
+- Don’t require Crossplane users to manage fine-grained RBAC access.
+- Don’t grant Crossplane providers broad RBAC access (e.g. to all other providers).
+
+## Proposal
+
+We propose introducing a new `Usage` type that can be used to declare usage
+relationships between Crossplane resources. This type will be defined by the
+Core Crossplane repository, and be available for use as part of a composition.
+
+The relations defined by this type will be enforced by an admission webhook that
+will be running as part of the Crossplane core. Similar to the upstreams [liens
+proposal], the webhook will reject deletions of resources that are in use by
+other resources with a ["409 Conflict" error code].
```

Comment:
We briefly talked about that in Istanbul: this is a misuse of the webhook API. A 409 is not considered as a non-admission, but as a failure of the webhook itself. The only valid response is through: https://github.com/kubernetes/kubernetes/blob/9e0569f2ed3934060fabe51be4e15232bbea3877/pkg/apis/admission/types.go#L124.

In systems like OpenShift and maybe also other Kube distributions and/or cloud providers, there will be monitoring alerts in place that will consider the 409 as a failure and warn the user, and at worst even disable the webhook or backoff in order to keep the cluster stable.

Reply:
Nvmd. The code snipped below shows what you actually do: you set 409 in the admission response. Let's clarify that here:
```suggestion
other resources with a ["409 Conflict" error code] as `metav1.Status` as part of the admission response (not the actual http code of the admission request).
```

Reply:
Thanks, updated accordingly.